drgn/helpers/linux/slab.py

```diff
@@ -53,3 +55,163 @@ def print_slab_caches(prog: Program) -> None:
         print(f"{name} ({s.type_.type_name()})0x{s.value_():x}")
+
+
+def obtain_freelist_slub(page_or_slab: Object, offset: int) -> Set[Object]:
+    """slub freelist is a pointer to the first free object"""
+    freelist_set = set()
+    freelist = page_or_slab.freelist.read_()
+
+    while freelist:
+        freelist_set.add(freelist.value_())
+        freelist = cast("void **", freelist + offset)[0].read_()
+    return freelist_set
+
+
+def obtain_freelist_slab(
+    prog: Program, page_or_slab: Object, freelist_index_size: int
+) -> Set[int]:
+    """slab freelist an array of free object indices"""
+    freelist_set = set()
+    freelist = page_or_slab.freelist.read_()
+
+    if freelist:
+        freelist_set.add(
+            int.from_bytes(prog.read(freelist, freelist_index_size), "big")
+        )
+        index_bytes = b""
+        index = 1
+        while index_bytes != b"\0":
+            index_bytes = prog.read(freelist + index, freelist_index_size)
+            freelist_set.add(int.from_bytes(index_bytes, "big"))
+            index += 1
+    return freelist_set
+
+
+def for_each_slab_page(prog: Program) -> Iterator[Object]:
+    """
+    Iterate over all slab pages
+    :return: Iterator of ``struct page *`` objects.
+    """
+    PGSlab = 1 << prog.constant("PG_slab")
+    for p in for_each_page(prog):
+        try:
+            if p.flags.value_() & PGSlab:
+                yield p
+        except FaultError:
+            pass
+
+
+def _get_slab_objects_slab_as_type(
+    prog: Program,
+    slab_cache: Object,
+    obj_type: Type,
+    slab_type: Type,
+    page_objects_func: Callable[[Program, Object, Object, Type], List[Object]],
+) -> List[Object]:
+    ret = []
+    for page in for_each_slab_page(prog):
+        slab = cast(slab_type, page)
+        if slab.slab_cache == slab_cache:
+            ret.extend(page_objects_func(prog, slab_cache, slab, obj_type))
+    return ret
+
+
+def _get_slab_objects_legacy(
+    prog: Program,
+    slab_cache: Object,
+    obj_type: Type,
+    page_objects_func: Callable[[Program, Object, Object, Type], List[Object]],
+) -> List[Object]:
+    ret = []
+    for page in for_each_slab_page(prog):
+        if page.slab_cache == slab_cache:
+            ret.extend(page_objects_func(prog, slab_cache, page, obj_type))
+    return ret
+
+
+def get_allocated_slab_objects(
+    prog: Program, slab_cache: Object, obj_type: Type
+) -> List[Object]:
+    """
+    Get the list of all allocated objects of a given slab and type
+
+    >>> dentry_slab = find_slab_cache(prog, "dentry")
+    >>> dentry_type = prog.type("struct dentry")
+    >>> get_allocated_slab_objects(prog, dentry_slab, dentry_type)
+    [Object(prog, 'struct dentry', address=0xffff888c8417de00), Object(prog, 'struct dentry', address=0xffff888c8417dec0)]
+
+    :param slab_cache: ``struct kmem_cache``
+    :param obj_type: type of object contained in the slab
+    """
```

Comment:
Helpers that take an `Object` shouldn't take a `Program` since you can get it from the `Object`.

```suggestion
def get_allocated_slab_objects(slab_cache: Object, obj_type: Type) -> List[Object]:
    """
    Get the list of all allocated objects of a given slab and type

    >>> dentry_slab = find_slab_cache(prog, "dentry")
    >>> dentry_type = prog.type("struct dentry")
    >>> get_allocated_slab_objects(prog, dentry_slab, dentry_type)
    [Object(prog, 'struct dentry', address=0xffff888c8417de00), Object(prog, 'struct dentry', address=0xffff888c8417dec0)]

    :param slab_cache: ``struct kmem_cache``
    :param obj_type: type of object contained in the slab
    """
    prog = slab_cache.prog_
```